website/blog/2025-03-25-velox-primer-part-2.mdx

```diff
@@ -0,0 +1,198 @@
+one query.
+
+Let's now look more closely at what each worker does at query setup time.
+
+## Task Setup
+
+In Prestissimo, the message that sets up a Task in a worker is called Task
+Update. A Task Update has the following information: the plan, configuration
+settings, and an optional list of splits. Splits are further qualified by what
+plan node they are intended for, and whether more splits for the recipient plan
+node and split group will be coming.  
+
+Since split generation involves enumerating files from storage (so they may
+take a while), Presto allows splits to be sent to workers asynchronously, such
+that the generation of spits can run in parallel with the execution of the
+first splits. Therefore, the first task update specifies the plan and the
+config. Subsequent ones only add more splits.
+
+Besides the plan, the coordinator provides configs as maps from a string key to
+a string value, both top level and connector level. The connector configs have
+settings for each connector; connectors are used by table scan and table writer
+to deal with storage and file formats. These configs and other information,
+like thread pools, top level memory pool etc. are handed to the Task in a
+QueryCtx object. See core/QueryCtx.h in the Velox repo for details.
+
+## From Plans to Drivers and Operators
+
+Once the Velox Task is created, a TaskManager hands it splits to work on. This
+is done with Task::addSplit(), and can be done after the Task has started
+executing. See velox/exec/Task.h for details.
+
+Let's zoom into what happens at Task creation: A PlanNode tree specifying what
+the Task does is given to the Task as part of a PlanFragment. The most
+important step done at Task creation is splitting the plan tree into pipelines.
+Each pipeline then gets a DriverFactory, which is the factory class that makes
+Drivers for the pipeline. The Drivers, in their turn, contain the Operators
+that do the work of running the query. The DriverFactories are made in
+LocalPlanner.cpp. See LocalPlanner::plan for details.
+
+<figure>
+    <img src="/img/velox-primer-part2-2.png" height= "100%" width="100%"/>
+</figure>
+
+Following the execution model known as Volcano, the plan is represented by an
+operator tree where each node consumes the output of its child operators, and
+returns output to the parent operator. The root node is typically a
+PartitionedOutputNode or a TableWriteNode. The leaf nodes are either
+TableScanNode, ExchangeNode or ValuesNode (used for query literals). The full
+set of Velox PlanNode can be found at velox/core/PlanNode.h. 
+
+The PlanNodes mostly correspond to Operators. PlanNodes are not executable as
+such; they are only a structure describing how to make Drivers and Operators,
+which do the actual execution. If the tree of nodes has a single branch, then
+the plan is a single pipeline. If it has nodes with more than one child
+(input), then the second input of the node becomes its own pipeline.  
```

Comment:
"becomes its own pipeline" it's a bit ambigious ( though I understand it's like hashjoin?)